### DISCUSSION TITLE
Request an Range of a object causes very high IO load

### DISCUSSION THREAD

**Discussion by hobbut** [minio/minio#16335] [Q&A]
## NOTE

I'm experiencing very high I/O load when i'm requesting more than 1.1MB chunk from the object.

## Expected Behavior
I/O load shouldn't be so high and the amount data read should be closer to the amount data served to user.

## Current Behavior
For 10000 requests of 1.1MB each (11GB data) For 11GB data received minio read from disks 235GB of data.
x21 read amplification
![image](https://user-images.githubusercontent.com/563097/209820897-e2abef1d-55eb-4a44-9321-f0906c468ae5.png)

## Steps to Reproduce (for bugs)

1. Upload 1GiB object to bucket
2. Run ab with Range header -H"Range: bytes=1-1100000" `ab -c32 -n10000 -H"Range: bytes=1-1100000" http://minio.test.server/test-bucket/1G`
3. Monitor IO on minio nodes. I used iotop `iotop -a -P -t | grep minio` to see accumulated I/O for minio process on each server

## Your Environment

* 4 servers with 8x1TB SSD disks. Minio running in container on each server
* minio version DEVELOPMENT.2022-12-07T00-56-37Z (commit-id=04ae9058ed5b135c27a3662b1153e129e5cf1483)
Runtime: go1.18.9 linux/amd64
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Copyright: 2015-2022 MinIO, Inc.
* `Linux as11 5.15.0-56-generic #62-Ubuntu SMP Tue Nov 22 19:54:14 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by harshavardhana** [comment]
@hobbut this is normal, MinIO is an erasure coded system and there is always more to read to erasure decode.

You are asking for a range many concurrent times and many times over. Each of these reads() do need an X amount of shards to fill-fill the request.

There is nothing surprising here. Also why are you compiling your own binaries please use officially published ones.

**Comment by harshavardhana** [comment]
To know all the calls that happen on the disk do 

```
mc admin trace -a --funcname os* --funcname storage* alias/
```

**Comment by hobbut** [comment]
@harshavardhana I also tried with the officially published binary. The results are same

```
minio version RELEASE.2022-12-12T19-27-27Z (commit-id=a469e6768df4d5d2cb340749fa58e4721a7dee96)
Runtime: go1.19.4 linux/amd64
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Copyright: 2015-2022 MinIO, Inc.
```

There is not problem with `Range: bytes=1-1048574` 2.9GB data read on each node
But starting with `Range: bytes=1-1048575` there is the problem. 58.75GB IO on each node for the same benchmark

`ab -c32 -n10000 -H"Range: bytes=1-1048574" http://server/test-bucket/1G_nochunks/1G`
![image](https://user-images.githubusercontent.com/563097/209921345-a03a72eb-24df-4e37-bd75-9a9b9c6d475b.png)

`ab -c32 -n10000 -H"Range: bytes=1-1048575" http://server/test-bucket/1G_nochunks/1G`
![image](https://user-images.githubusercontent.com/563097/209921145-da79086e-53c9-483a-9044-65f25242b63c.png)

**Comment by klauspost** [comment]
@hobbut Could you execute a single call of each and capture the output from `mc admin trace -call=s3 -call=internal -call=storage -call=os --json ALIAS`

Are you running encryption and/or have enabled compression?

**Reply by hobbut** [reply]
I don't have compression or encryption enabled.
There is log for each call

`Range: bytes=1-1048574`
[range_1_1048574.log](https://github.com/minio/minio/files/10327895/range_1_1048574.log)


`Range: bytes=1-1048575`
[range_1_1048575.log](https://github.com/minio/minio/files/10327896/range_1_1048575.log)

**Reply by harshavardhana** [reply]
> I don't have compression or encryption enabled. There is log for each call
> 
> `Range: bytes=1-1048574` [range_1_1048574.log](https://github.com/minio/minio/files/10327895/range_1_1048574.log)
> 
> `Range: bytes=1-1048575` [range_1_1048575.log](https://github.com/minio/minio/files/10327896/range_1_1048575.log)

Are you sure you captured everything?

```
wc -l ~/Downloads/range_1_1048574.log
99 /home/harsha/Downloads/range_1_1048574.log
```

```
wc -l ~/Downloads/range_1_1048575.log
99 /home/harsha/Downloads/range_1_1048575.log
```

With this result what you are saying doesn't make sense, it looks like some kind of wrong reporting by iotop.

**Reply by klauspost** [reply]
`grep -E ReadFileStream range*.log | grep -E callStats` shows first request requesting 87414 from each disk to get one block and second one requesting 174828 bytes. This is expected, since data is stored in 1MB blocks and needs a full block to read.

Following this further down, this makes a difference in whether [SmallFile is set](https://github.com/minio/minio/blob/2fa35def2c9a2c0a3620d9b46693b0ca49598631/cmd/xl-storage.go#L1803) for reads.

This will select a [bigger buffer](https://github.com/minio/minio/blob/97376f6e8f36007496716c3e5fe89771a7f48cb3/internal/ioutil/odirect_reader.go#L84) for reading, effectively reading 2MB instead of 32KB per read. This reduces syscalls and provides smoother streaming.

It could be agued that a threshold of 1MB for this behavior is a bit more reasonable than 128KB, though it will only really apply for this specific interval.

**Reply by klauspost** [reply]
Change proposed: https://github.com/minio/minio/pull/16338

**Comment by harshavardhana** [comment]
As @klauspost already asked we need to see your setup is it an encrypted blob or compressed?
